lib/quby/answers/entities/table_backend/disk_table.rb

```diff
@@ -0,0 +1,111 @@
+    #  a varied length non-float string array, to signify a set of categorical options (like 'male', 'female', 'other')
+    #    this will be returned as is
+    #  an empty array, which returns a range that will accept any value on an 'include?' call
+    def self.parse_range(range)
+      maybe_float_start = parse_float(range.first)
+      if maybe_float_start.present?
+        maybe_float_end = parse_float(range[1])
+        if maybe_float_end.present? && range.length == 2
+          return (maybe_float_start...maybe_float_end)
+        else
+          fail "#{range} has a float start but not a float end"
+        end
+      else # categorical range ([strings] that need no parsing), or empty array (which means 'accept anything')
+        if range.length >= 1
+          return range
+        else
+          return Quby::Answers::Entities::TableBackend::TableDimension::AcceptsAnythingRange.instance
+        end
+      end
+    end
+
+    def self.parse_float(maybe_float_string)
+      maybe_float = Float(maybe_float_string) rescue nil # can you strictly parse without using exceptions?
+      case maybe_float_string
+      when 'minfinity'
+        maybe_float = -Float::INFINITY
+      when 'infinity'
+        maybe_float = Float::INFINITY
+      end
+
+      maybe_float
+    end
+
+    def initialize(key)
+      path = Pathname.new self.class.disk_table_root.join(key)
+      fail "path #{path} outside of jail" unless self.class.validate_jailed_path(path)
+
+      @root_dimensions = dimensions_from_children(path)
+    end
+
+    def lookup(parameters)
+      result = nil
+      @root_dimensions.find do |dimension|
+        result = dimension.lookup(parameters)
+      end
+      result
+    end
+
+    def dimensions_from_children(directory_path)
+      children = Pathname.new(directory_path).children
+
+      # directory names are of the form 'dimensionname_rangestart_rangeend',
+      # files use the whole name as the dimension name, so we can use underscores in leaf dimension names
+      if children.all?(&:directory?)
+        dimensions_from_directories children
+      else
+        dimensions_from_files children
+      end
+    end
+
+    def dimensions_from_directories(directory_names)
+      dimensions = directory_names.group_by do |child|
+        child.expand_path.basename.to_s.split('_').first
+      end
+
+      dimensions.map do |dimension_key, pathnames|
+        ranges = pathnames.map do |pathname|
+          range_key = self.class.parse_range pathname.expand_path.basename.to_s.split('_')[1..-1]
+          range_value = dimensions_from_children(pathname)
+
+          [range_key, range_value]
+        end.to_h
+
+        TableDimension.new(dimension_key, ranges)
+      end
+    end
+
+    def dimensions_from_files(file_names)
+      file_names.map do |file_name|
+        rows = CSV.read(file_name)
+        ranges = rows.each_with_index.map do |(key, value), index|
+          next_key, _ = rows[index + 1]
```

Comment:
```
next_key = if (index + 1) < rows.length
             rows[index + 1].first
           else
             Float::INFINITY
```
is leesbaarder. Inline ifs zijn alleen leesbaar als guards. Guards bovenin als input validatie en een guard na een stap als je iets eigenlijk altijd wil doen, behalve op 2e kerstdag.